lms/djangoapps/mobile_api/middleware.py

```diff
@@ -0,0 +1,108 @@
+"""
+Middleware for Mobile APIs
+"""
+from datetime import datetime
+from django.core.cache import cache
+from django.http import HttpResponse
+from pytz import UTC
+from mobile_api.mobile_platform import MobilePlatform
+from mobile_api.models import AppVersionConfig
+from mobile_api.utils import parsed_version
+from openedx.core.lib.mobile_utils import is_request_from_mobile_app
+import request_cache
+
+
+class AppVersionUpgrade(object):
+    """
+    Middleware class to keep track of mobile application version being used
+    """
+    LATEST_VERSION_HEADER = 'EDX-APP-LATEST-VERSION'
+    LAST_SUPPORTED_DATE_HEADER = 'EDX-AAP-VERSION-LAST-SUPPORTED-DATE'
+    NO_LAST_SUPPORTED_DATE = 'NO_LAST_SUPPORTED_DATE'
+    NO_LATEST_VERSION = 'NO_LATEST_VERSION'
+    USER_APP_VERSION = 'USER_APP_VERSION'
+    REQUEST_CACHE_NAME = 'app-version-info'
+    CACHE_TIMEOUT = 3600  # seconds
+
+    def process_request(self, request):
+        """
+        raises HTTP Upgrade Require error if request is from mobile native app and
+        user app version is no longer supported
+        """
+        version_data = self._get_version_info(request)
+        if version_data:
+            last_supported_date = version_data[self.LAST_SUPPORTED_DATE_HEADER]
+            if last_supported_date != self.NO_LAST_SUPPORTED_DATE:
+                if datetime.now().replace(tzinfo=UTC) > last_supported_date:
+                    return HttpResponse(status=426)  # Http status 426; Update Required
+
+    def process_response(self, __, response):
+        """
+        If request is from mobile native app, then add headers to response;
+        1. EDX-APP-LATEST-VERSION; if user app version < latest available version
+        2. EDX-APP-UPGRADE-DATE; if user app version < min supported version and timestamp < expiry of that version
+        """
+        request_cache_dict = request_cache.get_cache(self.REQUEST_CACHE_NAME)
+        if request_cache_dict:
+            last_supported_date = request_cache_dict[self.LAST_SUPPORTED_DATE_HEADER]
+            if last_supported_date != self.NO_LAST_SUPPORTED_DATE:
+                response[self.LAST_SUPPORTED_DATE_HEADER] = last_supported_date.isoformat()
+            latest_version = request_cache_dict[self.LATEST_VERSION_HEADER]
+            user_app_version = request_cache_dict[self.USER_APP_VERSION]
+            if (latest_version != self.NO_LATEST_VERSION and
+                    parsed_version(user_app_version) < parsed_version(latest_version)):
+                response[self.LATEST_VERSION_HEADER] = latest_version
+        return response
```

Comment:
Future: Not a requirement for now, but in the future, consider the following for better code readability in this top-level method:

- Create a small class to encapsulate the `Date` checking and manipulations in this method.  Then, for example, instead of doing `if last_supported_date != self.NO_LAST_SUPPORTED_DATE` you can just do `if not last_supported_date`.
- Similarly, create a small class to encapsulate the `request_cache` access and manipulations.
- In the end, it would be great if the underlying implementations were encapsulated and the top-level method implementation would be closer to its semantics.  For example, this method would be re-written as:

```
if self._request_from_app():
    self._add_supported_date_if_needed(response)
    self._add_latest_version_if_needed(response)
```